lib/notifications/display.dart

```diff
@@ -39,6 +57,104 @@ class NotificationChannelManager {
+  }) {
+    const packageName = 'com.zulip.flutter'; // TODO(#407)
+
+    // URL scheme for Android resource url.
+    // See: https://developer.android.com/reference/android/content/ContentResolver#SCHEME_ANDROID_RESOURCE
+    const schemeAndroidResource = 'android.resource';
+
+    return Uri(
+      scheme: schemeAndroidResource,
+      host: packageName,
+      pathSegments: <String>[resourceTypeName, resourceEntryName],
+    );
+  }
+
+  /// Prepare our notification sounds; return a URL for our default sound.
+  ///
+  /// Where possible, this copies each of our notification sounds into shared storage
+  /// so that the user can choose between them in the system notification settings.
+  ///
+  /// Returns a URL for our default notification sound: either in shared storage
+  /// if we successfully copied it there, or else as our internal resource file.
+  static Future<String> _ensureInitNotificationSounds() async {
+    String defaultSoundUrl = _resourceUrlFromName(
+      resourceTypeName: 'raw',
+      resourceEntryName: kDefaultNotificationSound.resourceName).toString();
+
+    final shouldUseResourceFile = switch (await ZulipBinding.instance.deviceInfo) {
+      // Before Android 10 Q, we don't attempt to put the sounds in shared media storage.
+      // Just use the resource file directly.
+      // TODO(android-sdk-29): Simplify this away.
+      AndroidDeviceInfo(:var sdkInt) => sdkInt <= 28,
+      _                              => true,
+    };
+    if (shouldUseResourceFile) return defaultSoundUrl;
+
+    // First, look to see what notification sounds we've already stored,
+    // and check against our list of sounds we have.
+
+    final soundsToAdd = NotificationSound.values.toList();
+    final storedSounds = await _androidHost.listStoredSoundsInNotificationsDirectory();
+    for (final storedSound in storedSounds) {
+      assert(storedSound != null); // TODO(#942)
+
+      // If the file is one we put there, and has the name we give to our
+      // default sound, then use it as the default sound.
+      if (storedSound!.fileName == kDefaultNotificationSound.fileDisplayName
+          && storedSound.isOwner) {
+        defaultSoundUrl = storedSound.contentUrl;
+      }
+
+      // If it has the name of any of our sounds, then don't try to add
+      // that sound.  This applies even if we didn't put it there: the
+      // name is taken, so if we tried adding it anyway it'd get some
+      // other name (like "Zulip - Chime (1).m4a", with " (1)" added).
+      // Which means the *next* launch would try to add it again ad infinitum.
+      // We could avoid this given some other way to uniquely identify the
+      // file, but haven't found an obvious one.
+      //
+      // This does mean it's possible the file isn't the one we would have
+      // put there... but it probably is, just from a debug vs. release build
+      // of the app (because those have different package names).  And anyway,
```

Comment:
```suggestion
      // of the app (because those may have different package names).  And anyway,
```

Currently zulip-flutter differs from zulip-mobile in this respect. We might change that in the future, though, so the remark can remain, just needs this hedge.